### PR TITLE
Roman to int

### PR DESCRIPTION
**Pull request by Mushtary**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Okazia1715 — CHANGES_REQUESTED** [review]
Hi! I think there are some changes to be made. I wrote the details in the comments

File: solutions/Roman_to_Int.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    # Check for invalid patterns
+    for pattern in invalid_patterns:
+        if pattern in s:
+            raise ValueError(f"Invalid Roman numeral sequence: {s}")
+
+    total = 0  # Initialize total value
+    prev_value = 0  # Track the value of the previous numeral
+
+    # Loop through the string from right to left
```
[the remark this answers is not in the document]
Comment (Mushtary) [reply solutions/Roman_to_Int.py:44]: We loop through the Roman numeral from right to left because Roman numerals are usually written in decreasing order. However, sometimes a smaller numeral comes before a larger one to show subtraction (like IV for 4).

If the current numeral is smaller than the previous one, we subtract it (like in IV or IX). Otherwise, we just add it to the total. After each step, we update the previous numeral for the next comparison.

File: solutions/Roman_to_Int.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    Convert a Roman numeral string to an integer.
+
+    Args:
+    s (str): The Roman numeral string, containing characters from "I", "V", "X", "L", "C", "D", and "M".
+
+    Returns:
+    int: The corresponding integer value of the Roman numeral.
+
+    Raises:
+    ValueError: If the input contains invalid Roman numeral characters or sequences.
```
Comment (Okazia1715) [inline solutions/Roman_to_Int.py:12]: If the function has two assertions, I think there should be two "Raises"

File: solutions/Roman_to_Int.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Okazia1715) [reply solutions/Roman_to_Int.py:44]: Thank you, now I get it!

File: solutions/tests/test_roman_to_int.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    def test_roman_to_int_MCMXCIV(self):
+        """Test Roman numeral 'MCMXCIV'."""
+        self.assertEqual(roman_to_int("MCMXCIV"), 1994)
+
+    def test_roman_to_int_invalid(self):
+        """Test invalid Roman numerals."""
+        with self.assertRaises(ValueError):
+            roman_to_int("IIII")  # Invalid Roman numeral
+        with self.assertRaises(ValueError):
+            roman_to_int("ABC")  # Non-Roman input
```
Comment (Okazia1715) [inline solutions/tests/test_roman_to_int.py:41]: According to the checklist, each unittest should have only one assertion

**Review by Okazia1715 — CHANGES_REQUESTED** [review]

File: solutions/Roman_to_Int.py (hunk 2 above)
Comment (Mushtary) [reply solutions/Roman_to_Int.py:12]: Thank you for pointing that out! The function raises a ValueError for two distinct cases:

If the input contains invalid Roman numeral characters.
If the input contains invalid Roman numeral sequences.
Both cases are explicitly checked, and separate raise statements are included to handle each scenario. Let me know if further clarification is needed!

**Review by Okazia1715 — APPROVED** [review]
Thank you for your thoughtful work! I really appreciate you considering my comments. Nice job!

**Comment by Mushtary** [comment]
Thank you, Alona, for the review and useful comments that greatly helped improve my code. I appreciate it!
